homeassistant/components/escea/climate.py

```diff
@@ -0,0 +1,222 @@
+"""Support for the Escea Fireplace."""
+from __future__ import annotations
+
+from collections.abc import Callable, Coroutine
+import logging
+
+from pescea import Controller
+
+from homeassistant.components.climate import ClimateEntity
+from homeassistant.components.climate.const import (
+    FAN_AUTO,
+    FAN_HIGH,
+    FAN_LOW,
+    HVAC_MODE_HEAT,
+    HVAC_MODE_OFF,
+    SUPPORT_FAN_MODE,
+    SUPPORT_TARGET_TEMPERATURE,
+)
+from homeassistant.const import ATTR_TEMPERATURE, PRECISION_WHOLE, TEMP_CELSIUS
+from homeassistant.core import HomeAssistant, callback
+from homeassistant.helpers.dispatcher import async_dispatcher_connect
+from homeassistant.helpers.typing import ConfigType
+
+from .const import (
+    DATA_DISCOVERY_SERVICE,
+    DISPATCH_CONTROLLER_DISCONNECTED,
+    DISPATCH_CONTROLLER_DISCOVERED,
+    DISPATCH_CONTROLLER_RECONNECTED,
+    DISPATCH_CONTROLLER_UPDATE,
+    ESCEA,
+    ICON,
+)
+
+_LOGGER = logging.getLogger(__name__)
+
+_ESCEA_FAN_TO_HA = {
+    Controller.Fan.FLAME_EFFECT: FAN_LOW,
+    Controller.Fan.FAN_BOOST: FAN_HIGH,
+    Controller.Fan.AUTO: FAN_AUTO,
+}
+
+
+async def async_setup_entry(
+    hass: HomeAssistant, config: ConfigType, async_add_entities: Callable
+):
+    """Initialize an Escea Controller."""
+    disco = hass.data[DATA_DISCOVERY_SERVICE]
+
+    @callback
+    def init_controller(ctrl: Controller):
+        """Register the controller device."""
+
+        _LOGGER.info("Controller UID=%s discovered", ctrl.device_uid)
+
+        device = ControllerDevice(ctrl)
+        async_add_entities([device])
+
+    # create any components not yet created
+    for controller in disco.pi_disco.controllers.values():
+        init_controller(controller)
+
+    # connect to register any further components
+    async_dispatcher_connect(hass, DISPATCH_CONTROLLER_DISCOVERED, init_controller)
```

Comment:
Sorry, I've struggled with this change and the ones below. (For the most part) I understand the structure I've used (which used iZone as the template) and seeing as it has been working well for me the past couple of months, I'd rather not make changes like these (that I'd struggle to debug).

Reply:
```suggestion
    config.async_on_unload(async_dispatcher_connect(hass, DISPATCH_CONTROLLER_DISCOVERED, init_controller))
```

Wrap this or it will leak when the integration is unloaded.

Also please rename `config` to `entry` or `config_entry` as its a config entry

Reply:
It looks like each device has its own ip address and there is no hub.  In this type of setup I recommend a config entry per device since we can use Home Assistant's internal retry logic to manage retrying to connect to the device and handling ip changes.  Additionally when a known device is offline, the UI can provide feedback on the integrations screen about why Home Assistant cannot connect to a specific device.  When all the devices are under a single config entry the integration has to re-invent many of these built-ins

`tplink`, `yeelight`, and `flux_led` all have enables on how to run discovery and use a single config entry per device as well as updating the ip address via discovery when the device changes ip.